### ISSUE TITLE
Feature request: ability to get a null/undefined

### ISSUE THREAD

**Issue by C-Saunders** [true-myth/true-myth#63]
Use case example - the ORM I'm using (Objection.js) wants a `null` or `undefined` to not update a field. In a situation where the field is optional (e.g. for an update), it would be nice to be able to do something like:

```ts
Person.patch({
  firstName: Maybe.of(request.firstName).unwrapOrNull()
})
```

In this case `unsafelyUnwrap()` would yield an `undefined`, which might work in this scenario, but it throws instead.

Thanks for taking a look and for the library (and New Rustacean)!

**Comment by chriskrycho** [comment]
Interesting proposal! @bmakuh would be curious to hear your thoughts on this as well, but at first blush I'm amenable, albeit probably with looooots of caveats in the docs about it.

(And you're welcome for New Rustacean! ❤️ )

**Comment by bmakuh** [comment]
If you're just using plain JS, then you could probably just do `Maybe.of(request.firstName).value`. It'll be undefined if you've got `Nothing`. If you're in TS, it's a little more verbose but you could do this:
```ts
const firstName = Maybe.of(request.firstName)
Person.patch({
  firstName: Maybe.isJust(firstName) ? firstName.value : null
})
```

`unwrapOrNull` is essentially syntactic sugar for that. I feel torn about adding this feature, because on the one hand there _are_ use cases for interop with libraries that use null/undefined in this kind of way, but on the other hand, the point of the library is to get away from using `null`. Obviously this is a sort of "escape hatch" on the edge of the library, but I don't know that people would see it that way. 

tl;dr: I think I actually prefer the painful verbosity of extracting a null because in the main it makes it hard to reach for what you know and makes you think in variant types rather than in nullability.

**Comment by flyiniggle** [comment]
This is a really interesting feature proposal to me. Like @bmakuh said, I know the whole point of a `Maybe` type is to get away from null values, but, practically speaking, the rest of the JS ecosystem makes explicit use of this language feature/mistake (depending on your perspective :)). I think this would be a useful feature when it comes to integration with other libraries and frameworks.

In my day to day work, a common manifestation of this is optionally rendering components in React based on the existence or value of a particular piece of state. If a given portion of some state exists, it’s easy to wrap that in a `Maybe`, map it to a component (often with intermediate calculations), and unwrap it to render the component. React expects the expression I’m trying to render to resolve to `null` if I don’t want the component to render, but of course I can't go around extracting null values from my `Maybe`. Obviously there are ways around this - for this specific case, I just made a component that renders null and use that as my default return. I think there’s a pretty strong argument that that’s a more ‘correct’ way of doing things, but honestly it would be convenient to not have to deal with that extra component, and I’ve had other situations where the solution wasn’t quite so pretty. I totally get the inclination to keep this inelegant so people will try to solve these problems with other type variants, but I also don’t want to make my code more verbose than necessary just because, at the end of the day, I’m still working with JS/TS and have to deal with null sometimes.

I’d be happy to make a quick PR for this, although I understand if this just doesn’t fit philosophically. I definitely agree that there should be caveats in the docs about using it. Maybe it could even be named `unsafelyUnwrapOrNull`?

**Comment by bmakuh** [comment]
@flyiniggle so this is super interesting to me. I also struggle with the interop question. Obviously if you're dealing with a language where _everything_ is built with variants in mind, then you never run into uncomfortable situations like you're describing. To me, I'd like to take a step back for a second into the broader question of design before debating this specific method. How do you build a library like this that creates a "pit of success"? We want to make it very easy to bring data into a type variant rather than dealing with raw nullability, but we don't want to make it trivial to bring it back out again because then you end up straddling the fence between normal JS ways of doing things and doing things in a type variant kind of way. For example, you could do this:

```tsx
function MyNeatComponent() {
  const firstName = Maybe.of(this.props.firstName)

  if (firstName.isNothing()) {
    return null
  }

  return (
    <div>{firstName.value}</div>
  )
}
```

But while I'm an FP nerd and I like modeling missing data with Maybe variants, if I'm intellectually honest, the above doesn't really warrant bringing in the size and complexity of an additional library when I'm basically just doing this:

```tsx
function MyNeatComponent() {
  const { firstName } = this.props

  if (!firstName) {
    return null
  }

  return (
    <div>{firstName.value}</div>
  )
}
```

It's a trivial example and doesn't reflect real-world data complexity, but now in a world where we have `?.` and `??` operators, it's honestly pretty trivial to handle nullability without having to litter your code with big null guard blocks everywhere. I still would prefer to handle uncertainty "at the edge of the system" so to speak so that I have data guarantees throughout the rest of the app, but what does that actually have to look like? If we were to design the next major version of True Myth with these new operators in mind, what value proposition would True Myth provide? 

Another way to say it is, how do we provide the niceties that monadic variants like Maybe and Result give you while also acknowledging that JavaScript is what it is and not trying to make it into something that it's not? To me, I find it to be a very nice experience to be able to `map`, `ap`, etc. and do data transformations on a piece of data without having to worry about whether it exists or not. 

I work with libraries like React and Sequelize and various native DOM APIs that assume "nothing" will be represented by `null` or `undefined`, so I would also like to have a really killer interop story that would sell itself. I'd just love to solve that at the design level rather than just growing the library organically by adding a method here and a method there; let's make sure the design "fits" with plenty of use cases. Thoughts?

**Comment by flyiniggle** [comment]
Yes, I completely agree that the `?.` and `??` operators make using `Maybe` for property access and default values pretty uninspiring. On the other hand, this was never really the primary draw for me. If I'm being honest, most of my co-workers who cared rather less for FP than I do never thought it was enough of a value add to warrant an extra library even before null coalescing etc., so I think it's a really good idea to question what the value of `Maybe` and `Result` is in the context of a language that doesn't conceptually include them.

To me, the value is mostly in the ability to write functions that can be pure morphisms between types, compose those together, and then apply them to situations that may involve `null`/`undefined` or exceptions - all without having the added burdens of dealing with that in every single function I write. In practice in my code, this usually ends up looking like:

1. Data comes in to some sort of interface - it may be `null` or some other type
2. Wrap the data in a Maybe. 
3. Map that value to some other type
4. Unwrap the value to pass to some API I don't control, or return it from the API.

The obvious problem here is that we have to deal with (often imperative) code that has no idea what a `Maybe` is and instead embraces `null`/`undefined`. I don't think there's much more to be done on the input side. On the other side, I sometimes do something like:

```
const effect = maybeAValue.match({
    Just: (x => doSomethingEffectful.bind(null, x)),
    Nothing: (x => () => undefined)
});

effect();
```

Of course, I think it's exactly this sort of things that makes people hate FP when they first see it, so I can't say I love it. I think an `Effect` or `Io` kind of thing is the usual way of mapping types to effects, so maybe that's the way to go. I think that's vaguely equivalent to what I'm describing with `match`, but with the added benefit of composabilty. I think this is discussed in another ticket somewhere.

Ok, so for this next bit, keep in mind that my understanding of category theory is mediocre at best. Please correct or completely disregard what follows, as appropriate.

To the best of my understanding, `Maybe` lets you take a function that would normally map one type to another (or the same type) and instead apply it to that original type _and_ `null`/`undefined`. The starting type is actually a set of the values that the type can represent; `null` is a singleton set. With `Maybe`, you get the disjoint union of the two sets. Similarly, `Result` gives you the disjoint union of any two sets, but often one type and the empty set (aka an exception). This is really great, except at the end of all the mapping you usually need to _do_ something with the result of those objects and morphisms. That means mapping them to effects or just taking them out of the wrapper. I think the former is something that it might be good for this library to support, but it doesn't cover all the use cases. Specifically, let's go back to React, or any other library that let's you return JSX from a function. With that kind of thing, you're letting the library handle the effects and all you actually want is the raw value. If `Maybe` represents a value that can be either whatever type you're using or the singleton set of `null`, it seems like having an "unwrap" function that returns `null` if `Maybe.isJust() === false` is "correct" in some ways.

I'm not sure all of that exposition actually suggests a clear path forward, but I'd love to hear your thoughts.

**Comment by chriskrycho** [comment]
Sorry for the long delay; I've been mulling on this off and on for the past couple months (and @bmakuh and I have chatted about it a bit in the background as well). I'll try to provide a more detailed response in the next week, but I'd summarize that the two constraints I see here are:

- We *should* provide reasonable ways to interoperate with existing JS APIs which do *not* understand the types we provide (which is basically all of them!), at all edges of the system. That means that we need both ways to map into our types when data is *entering* the system, and to map our types back out when data is *leaving* the system.
- We shouldn't, in addressing that first constraint, compromise the specific value propositions this *does* add in terms of type safety and ergonomics (though, as @bmakuh noted, the ergonomic value of True Myth is somewhat decreased for nullable types given `?.` and `??`; and my initial experiments with a version of True Myth where those were available were… *interesting*, let's say).

I don't have a concrete way forward here, and I'm not intrinsically opposed to adding an 'escape hatch' that makes those boundary interop modes better.

Elm's JSON decoding and encoding is an interesting point of analogy here: it makes you define explicitly how you want to transform your safe representation into the loose JS version. I wonder if we might do something similar, in particular by providing a *function* import—*not* a method on the classes—that would clearly be doing something for the interop cases, and therefore not *quite* as easy to reach for (because @bmakuh's concerns are legitimate there!). 🤔

**Comment by atrick-speedline** [comment]
How about changing the type on `unwrapOr` to this?

```ts
unwrapOr(this: Result<T, E>, defaultValue: U): T | U;
```

**Comment by chriskrycho** [comment]
Huh. That's a very interesting suggestion.

**Comment by bmakuh** [comment]
Sorry to take so long on this, but @atrick-speedline I actually think that's a pretty good idea. @chriskrycho if you're okay with this, I would say go ahead and file a PR and we'll review it!

**Comment by chriskrycho** [comment]
Yeah, I keep coming back to this over the last couple months mentally and I like it. 👍

**Comment by bmakuh** [comment]
Cool cool. @atrick-speedline if you're still interested in this feature, feel free to file the PR.

**Comment by chriskrycho** [comment]
Resolved via #69, which was published as v4.1.0!
